File: yogi-cpp/test/public_api_tests/operation_id_test.cc
/*
 * This file is part of the Yogi distribution https://github.com/yohummus/yogi.
 * Copyright (c) 2018 Johannes Bergmann.
 *
 * This library is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 3.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this library. If not, see <http://www.gnu.org/licenses/>.
 */

#include "../common.h"

#include <vector>
#include <type_traits>

class OperationIdTest : public testing::Test {
 protected:
  const yogi::OperationId oid_ = yogi::internal::MakeOperationId(123);
};

TEST_F(OperationIdTest, DefaultConstructor) {
  yogi::OperationId oid;
  EXPECT_FALSE(oid.IsValid());
  EXPECT_EQ(oid.Value(), 0);
}

TEST_F(OperationIdTest, Value) { EXPECT_EQ(oid_.Value(), 123); }

TEST_F(OperationIdTest, IsValid) {
  EXPECT_TRUE(oid_.IsValid());

  auto oid = yogi::internal::MakeOperationId(0);
  EXPECT_FALSE(oid.IsValid());

  oid = yogi::internal::MakeOperationId(-1);
  EXPECT_FALSE(oid.IsValid());
}

TEST_F(OperationIdTest, Copy) {
  yogi::OperationId oid;
  oid = oid_;
  EXPECT_EQ(oid.Value(), 123);
  EXPECT_TRUE(oid.IsValid());
}

File: yogi-dotnet/test/OperationIdTest.cs
/*
 * This file is part of the Yogi distribution https://github.com/yohummus/yogi.
 * Copyright (c) 2018 Johannes Bergmann.
 *
 * This library is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 3.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this library. If not, see <http://www.gnu.org/licenses/>.
 */

using System;
using Xunit;

namespace test
{
    public class OperationIdTest : TestCase
    {
        [Fact]
        public void DefaultValue()
        {
            var id = new Yogi.OperationId();
            Assert.Equal(id, new Yogi.OperationId(0));
        }

        [Fact]
        public void Constructor()
        {
            var id = new Yogi.OperationId(5);
            Assert.Equal(5, id.Value);
            Assert.True(id.IsValid);

            id = new Yogi.OperationId(-12);
            Assert.Equal(-12, id.Value);
            Assert.False(id.IsValid);

            id = new Yogi.OperationId(0);
            Assert.Equal(0, id.Value);
            Assert.False(id.IsValid);
        }

        [Fact]
        public void Comparison()
        {
            var id1 = new Yogi.OperationId(11);
            var id2 = new Yogi.OperationId(11);
            var id3 = new Yogi.OperationId(333);

            Assert.True(id1 == id2);
            Assert.False(id1 == id3);
            Assert.False(id1 == null);

            Assert.False(id1 != id2);
            Assert.True(id1 != id3);
            Assert.True(id1 != null);

            Assert.False(id1.Equals(new Exception()));
            Assert.False(id1.Equals(null));
            Assert.True(id1.Equals(id2));
            Assert.False(id1.Equals(id3));
        }
    }
}
